### ISSUE TITLE
Haxe 4.0.0-rc.2 breaks defold bundling.

### ISSUE THREAD

**Issue by jurmerlo** [hxdefold/hxdefold#13]
I was using haxe 3.4.7 and that works fine but updated to haxe 4.0.0-rc.2 to get better support in vscode. It works when I build in the editor, but when I try to make a bundle for MacOS or Linux I get
`/main.lua the module '/bit32.lua' can't be found`.
I haven't tested other platforms yet.

I checked why it happens and found this part of the generated code has changed:
Version 3.4.7:
```
pcall(require, 'bit')
if bit then
  _hx_bit = bit
elseif bit32 then
  local _hx_bit_raw = bit32
  _hx_bit = setmetatable({}, { __index = _hx_bit_raw });
  _hx_bit.bnot = function(...) return _hx_bit_clamp(_hx_bit_raw.bnot(...)) end;
  _hx_bit.bxor = function(...) return _hx_bit_clamp(_hx_bit_raw.bxor(...)) end;
else
  _G.error("Bitop library is missing.  Please install luabitop");
end
```

Version 4.0.0-rc2:
```
pcall(require, 'bit')
if bit then
  _hx_bit = bit
else
  local _hx_bit_raw = _G.require('bit32')
  _hx_bit = setmetatable({}, { __index = _hx_bit_raw });
  _hx_bit.bnot = function(...) return _hx_bit_clamp(_hx_bit_raw.bnot(...)) end;
  _hx_bit.bxor = function(...) return _hx_bit_clamp(_hx_bit_raw.bxor(...)) end;
end
```

I also tried the lastest nightly build of haxe and it generates the same code as 4.0.0-rc2.
Is there a solution or workaround for this?

**Comment by nadako** [comment]
@jdonaldson any comments?:)

**Comment by nadako** [comment]
> when I try to make a bundle for MacOS or Linux

This is actually interesting. Judging by the generated code, it checks for `bit` which should be present in defold (https://www.defold.com/ref/bit/), is it documented somewhere that it's not available for MacOS/Linux targets? Could it be a bug on Defold side?

**Comment by jdonaldson** [comment]
Yeah, ``bit`` should be there in defold.  I removed the thrown error because it was causing problems for some people in extremely limited lua installations.  I'm not sure what I could improve here TBH.

**Comment by jurmerlo** [comment]
I built a html5 bundle so I could easily see print in the console. The html5 and ios bundles also give the same error as above. If I take the whole else part out or just set ```local _hx_bit_raw = nil``` it builds fine. I added a print statement inside 
```
if bit then
  print('found bit')
  _hx_bit = bit
else
```
and this shows up in the console in the html5 build so it finds bit.

I think defold tries to  require bit32 when building a bundle even though it is not needed when running the code.

**Comment by nadako** [comment]
I just tried it. Looks like it is not even a run-time error, it's a Defold build error. I can't say for sure, but I suspect that defold bundling searches for all `require` calls in the code and executes them so it knows what to bundle, without caring about the condition branches.

It can be reproduced in pure Lua defold script like this:

```lua
function init(self)
	if false then
		_G.require("tralala")
	end
end
```

Interestingly, if you use `require` instead of `_G.require` it starts failing even with simple debug runs (F5 on my Windows machine).

**Comment by nadako** [comment]
I think one could just create an empty `bit32.lua` in the project to work this around, but I'd really love to hear some comments from the Defold devs here.

**Comment by jurmerlo** [comment]
Yeah an empty ```bit32.lua``` works. I'll use that for now thanks.

**Comment by nadako** [comment]
@jdonaldson I think on our side, we should move these hard-coded things into `.lua` files like we do for `_hx_tab_array` and friends, so people can override them in their classpaths. E.g. I could provide a version without `bit32` requirement directly in hxdefold.

**Comment by nadako** [comment]
I've created two issues
https://github.com/defold/editor2-issues/issues/2527 - let's see what Defold people have to say about this
https://github.com/HaxeFoundation/haxe/issues/8051 - we can make Haxe more flexible with these things

Leaving this issue open until we figure out what to do next. The empty-module workaround should do for now.

**Comment by jdonaldson** [comment]
> @jdonaldson I think on our side, we should move these hard-coded things into `.lua` files like we do for `_hx_tab_array` and friends, so people can override them in their classpaths. E.g. I could provide a version without `bit32` requirement directly in hxdefold.

I think I follow you, and this seems like we're coming up with a basic strategy for handling polyfills.  How does this mesh with the js target?  It would be nice to have some consistency in behavior for that.

**Comment by nadako** [comment]
There's not much polyfills in JS target nowadays, and the only stuff that is "hard-coded" in the generator is some helper functions for ES5 inheritance and such. For Lua we currently do [this](https://github.com/HaxeFoundation/haxe/blob/development/src/generators/genlua.ml#L1956) to include helper code for some stuff. Since it uses `Common.find_file`, I think it respects class path order, so it can be overriden in libs and custom class paths. We just need to move other hard-coded stuff into similar files.

**Comment by thejustinwalsh** [comment]
Any update on the final solution for this?

**Comment by nadako** [comment]
I have to look into this again when I get some time, I think if we up Haxe version requirement for hxdefold we can overshadow the bit stuff in the library.

**Comment by jdonaldson** [comment]
Let me know if you find any other nits, I'm running into these situations pretty often.  I believe it's easiest to just override the lua polyfills, but I keep thinking that there's a more elegant way of handling this.

**Comment by gerroon** [comment]
I have this issue with Haxe 4.3.3. Defold 1.6.3 and hxdefold example projects pretty much all like hxdefold-example-throwacrow-master I am trying to see if Haxe+Defold would be a good dev solution bso far I was not able to compile any of the provided examples.

Putting an empty bit32 file resolves the initial complaint but then it throws another one in `hxdefold-example-throwacrow-master`


![image](https://github.com/hxdefold/hxdefold/assets/8519469/dc9adaca-f9ec-4b90-af3c-fd2b44b64544)

**Comment by haath** [comment]
@gerroon 
During the transition to hxdefold v2 most example repos are outdated.

If you need a working example. the only one I've updated to v2 is: [hxdefold-bunnymark](https://github.com/hxdefold/hxdefold-bunnymark)
I just confirmed that it works fine with Defold 1.6.3

**Comment by gerroon** [comment]
@haath Thanks for the reply. It looks like haxelib installed `hxdefold: [1.5.0] git`  (based on the readme) The haxelib repo still installs 1.5. How do I grab the update? I am not sure if it is stable now.

**Comment by haath** [comment]
@gerroon v2 isn't merged into master yet so it's not on lib.haxe.org either. The only way to add it currently is to set hxdefold as a git dependency, and specifically the `v2` branch. You can see how in the bunnymark hxml:

```sh
--library hxdefold:git:https://github.com/hxdefold/hxdefold.git#v2
```

> It looks like haxelib installed...

It's not a good practice generally to rely on what haxelib decided to install for dependency management.
Better to specify what versions of everything you need (as above), and then let `haxelib install build.hxml` do the rest :wink:
